### ISSUE TITLE
Question about theming and CSS modules

### ISSUE THREAD

**Issue by bvaughn** [zippyui/react-date-picker#59]
Firstly, this is a discussion issue. Not a defect.

I'm the author of [react-virtualized](https://github.com/bvaughn/react-virtualized) and I'm currently working on extracting my styles into a separately-loaded CSS file like you do ([react-virtualized/issues/48](https://github.com/bvaughn/react-virtualized/issues/48)).

However I've hit a bit of a snag when testing my updated component in the popular [erikras/react-redux-universal-hot-example](https://github.com/erikras/react-redux-universal-hot-example) boilerplate: CSS files are imported as modules with scrambled names. I could inform people that they'll need to configure Webpack to `exclude` the theme CSS and import it without modules but that feels janky. (Who wants to add more complexity to their Webpack configurations anyway?)

Basically I wanted to touch base with you and see what your thoughts were on the subject. It seems like your approach just doesn't support CSS modules, but I don't see any issues reported against the date picker for it so maybe it hasn't turned out to be much of a problem?

Thanks in advance for any discussion. Feel free to close this issue at any point. :)

**Comment by radubrehar** [comment]
Thanks for pinging me in. I just responded https://github.com/bvaughn/react-virtualized/issues/48#issuecomment-170488846 but we can keep this open if you want, for further discussion on the topic.
Basically, people should not use css-modules to load packages/files from node_modules. Css modules should only be used for app-specific components/files. Otherwise, they risk throwing away light-years of css code/work that will simply not work with css-modules because of the generated classnames.

Let me know if you have any other questions.
By the way, nice job on react-virtualized. It seems we're doing some similar work, see our react-datagrid :) keep up the good work!

**Comment by bvaughn** [comment]
Thank you! And same to you! :)

I ended up trying something new with the react-virtualized release in an effort to simplify universal rendering. (Although I'm not going to lie- part of me still wonders if it's over-engineered and simple, static class names and a default prefix CSS file wouldn't have been better.)

Basically I use inline stiles but with the option to configure and override them on a per-instance or per-component basis. Styles get auto-prefixed at runtime (using [inline-style-prefixer](https://github.com/rofrischmann/inline-style-prefixer)) before being applied.

**Comment by radubrehar** [comment]
Inline style prefixing is problematic. I know since I've implemented https://www.npmjs.com/package/react-style-normalizer (used by  http://reapp.io/) for the same reason, to accomplish universal rendering. But unless you apply all prefixes while rendering on the server side & the client side, it's really more of a headache, since when attaching on the client, React will complain that the initial rendering does not match the client-rendering.
Ideally you should only apply the prefixes of the current browser, but in this way, the server-side rendering is out of question, for the reason mentioned above.

Also I don't think people will succeed to implement a prefixing solution for inline styles as solid as the autoprefixer is. So for the moment, I recommend sticking to plain'ole css for highly-reusable components.

**Comment by bvaughn** [comment]
Prefixed styles aren't really necessary for server-side rendering. I assumed it would be okay to render non-prefixed styles on the server and then do prefixing in `componentDidMount` (when a browser could actually be detected). I assume as you say that this isn't as good as autoprefixer- (it's data/config file is huge)- but for the basics I thought it seemed like a reasonable approach.

What am I missing? :)

**Comment by radubrehar** [comment]
I assume you are trying to achieve universal rendering, without any performance issues/extra rendererings. So on ReactDOM.render, to render without issues, you should render content identically with what the server already rendered. Otherwise, if there are some differences, you will have warnings, AFAIK.
Doing prefixing on client-side only in didMount is more of a hack in my opinion and it doesn't scale.

**Comment by bvaughn** [comment]
I've come around to this point of view as well, thanks in part to this
discussion. Released v4 of react-virtualized last night which has all
styles extracted into a prefixed styles.css file.

Thanks for the useful back and forth. Cheers :)

**Comment by radubrehar** [comment]
That looks much better imo. Keep up the good work! Glad to have helped!
